src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp

```diff
@@ -0,0 +1,487 @@
+    this->qNoise.resize(this->state.size(), this->state.size());
+
+    this->timeTag = CurrentSimNanos*NANO2SEC;
+    this->numberSigmaPoints = this->state.size()*2+1;
+    this->dt = 0;
+
+    /*! - Ensure that all internal filter matrices are zeroed*/
+    this->obs.setZero(this->state.size());
+    this->wM.setZero(this->numberSigmaPoints);
+    this->wC.setZero(this->numberSigmaPoints);
+    this->sBar.setZero(this->state.size(), this->state.size());
+    this->sigmaPoints.setZero(this->state.size(), this->numberSigmaPoints);
+    this->sQnoise.setZero(this->state.size(), this->state.size());
+
+    /*! - Set lambda/gamma to standard value for unscented kalman filters */
+    this->lambda = this->state.size()*(this->alpha*this->alpha - 1);
+    this->eta = sqrt(this->state.size() + this->lambda);
+
+    /*! - Set the wM/wC vectors to standard values for unscented kalman filters*/
+    this->wM(0) = this->lambda / (this->state.size() + this->lambda);
+    this->wC(0) = this->lambda / (this->state.size() + this->lambda) + (1 - this->alpha*this->alpha + this->beta);
+    for (int i = 1; i < this->numberSigmaPoints; i++)
+    {
+        this->wM(i) = 1.0 / (2.0 * (this->state.size() + this->lambda));
+        this->wC(i) = this->wM(i);
+    }
+
+    /*! - Perform cholesky decompositions of covariance and noise */
+    this->sBar = FlybyODuKF::choleskyDecomposition(this->covar);
+    this->sQnoise = FlybyODuKF::choleskyDecomposition(this->qNoise * 1E-6);
+}
+
+/*! This method takes the relative position measurements and outputs an estimate of the
+ spacecraft states in the inertial frame.
+ @return void
+ @param configData The configuration data associated with the OD filter
+ @param callTime The clock time at which the function was called (nanoseconds)
+ @param moduleId The ID associated with the configData
+ */
+void FlybyODuKF::UpdateState(uint64_t CurrentSimNanos)
+{
+    double newTimeTag;  /* [s] Local Time-tag variable*/
+    bool computePostFits = false;
+
+    /*! - If the time tag from the measured data is new compared to previous step,
+     propagate and update the filter*/
+    this->opNavHeadingBuffer = this->opNavHeadingMsg();
+    newTimeTag = this->opNavHeadingBuffer.timeTag * NANO2SEC;
+    if(newTimeTag >= this->timeTag && this->opNavHeadingBuffer.valid)
+    {
+        TimeUpdate(newTimeTag);
+        MeasurementUpdate();
+        computePostFits = true;
+    }
+    /*! - If current clock time is further ahead than the measured time, then
+     propagate to this current time-step*/
+    newTimeTag = CurrentSimNanos*NANO2SEC;
+    if(newTimeTag >= this->timeTag)
+    {
+        TimeUpdate(newTimeTag);
+    }
+
+    this->opNavFilterMsgBuffer = this->opNavFilterMsg.zeroMsgPayload;
+    this->navTransOutMsgBuffer = this->navTransOutMsg.zeroMsgPayload;
+
+    /*! - The post fits are y - ybar if a measurement was read, if observations are zero, do not compute post fit residuals*/
+    if(computePostFits){
```

Comment:
:large_orange_diamond: This could be factored out into a `computePostFitResiduals` method.

Reply:
Done